scheduler/cloudformation/migration.go

```diff
@@ -0,0 +1,172 @@
+	// to stabilize before swapping over to it. 5-10 minutes is probably
+	// enough for the load balancer to starting being resolvable.
+	WaitTime time.Duration
+
+	// The scheduler that we want to migrate to.
+	cloudformation *Scheduler
+
+	// The scheduler we're migrating from.
+	ecs *ecs.Scheduler
+
+	db *sql.DB
+}
+
+// NewMigrationScheduler returns a new MigrationSchedeuler instance.
+func NewMigrationScheduler(db *sql.DB, c *Scheduler, e *ecs.Scheduler) *MigrationScheduler {
+	return &MigrationScheduler{
+		WaitTime:       10 * time.Minute,
+		db:             db,
+		cloudformation: c,
+		ecs:            e,
+	}
+}
+
+// Backend returns the scheduling backend to use for the given app.
+func (s *MigrationScheduler) Backend(appID string) (scheduler.Scheduler, error) {
+	backend, err := s.backend(appID)
+	if err != nil {
+		return nil, err
+	}
+
+	switch backend {
+	case "ecs":
+		return s.ecs, nil
+	case "cloudformation":
+		return s.cloudformation, nil
+	case "migrate":
+		return nil, fmt.Errorf("%s is currently being migrated to a CloudFormation stack. Sit tight...", appID)
+	default:
+		return nil, fmt.Errorf("unexpected scheduling backend encountered: %s", backend)
+	}
+}
+
+// backend returns the name of the backend to use for operations.
+func (s *MigrationScheduler) backend(appID string) (string, error) {
+	var name string
+	err := s.db.QueryRow(`SELECT backend FROM scheduler_migration WHERE app_id = $1`, appID).Scan(&name)
+
+	// For newly created apps.
+	if err == sql.ErrNoRows {
+		return "cloudformation", nil
+	}
+
+	return name, err
+}
+
+// Migrate prepares this app to be migrated to the cloudformation backend. The
+// next time the app is deployed, it will be deployed to using the
+// cloudformation backend, then the old resources will be removed.
+func (s *MigrationScheduler) Prepare(appID string) error {
+	_, err := s.db.Exec(`UPDATE scheduler_migration SET backend = 'migrate' WHERE app_id = $1`, appID)
+	return err
+}
+
+func (s *MigrationScheduler) Submit(ctx context.Context, app *scheduler.App) error {
+	backend, err := s.backend(app.ID)
+	if err != nil {
+		return err
+	}
+
+	if backend == "migrate" {
+		return s.Migrate(ctx, app)
+	}
+
+	b, err := s.Backend(app.ID)
+	if err != nil {
+		return err
+	}
+	return b.Submit(ctx, app)
+}
+
+// Migrate submits the app to the CloudFormation scheduler, waits for the stack
+// to successfully create, then removes the old API managed resources using the
+// ECS scheduler.
+func (s *MigrationScheduler) Migrate(ctx context.Context, app *scheduler.App) error {
+	// Submit to cloudformation and wait for it to complete successfully.
+	// Don't make any DNS changes.
+	if err := s.cloudformation.SubmitWithOptions(ctx, app, SubmitOptions{
+		Wait:  true,
+		NoDNS: true,
+	}); err != nil {
+		return err
+	}
+
+	// Wait for some time for everything to start running.
+	<-time.After(s.WaitTime)
+
+	// Remove the old AWS resources.
+	if err := s.ecs.Remove(ctx, app.ID); err != nil {
```

Comment:
Yeah, right now that's hidden, but we could expose that on the ecs scheduler. Might be worth it.

I was thinking, we could also make the migration a 2-3 step process:

1. The new stack is created, and DNS is swapped over to it.
2. DNS is swapped.
3. The old resources are removed.

For high load services, this might be ideal, because it gives you time to pre-warm ELB's.

Moving through each step could also potentially be controlled by just setting environment variables with `emp set`, like:

1. `emp set EMPIRE_SCHEDULER_MIGRATION=step1`.
  * Wait for the stack to be created.
  * Ask AWS to pre-warm the ELB.
2. `emp set EMPIRE_SCHEDULER_MIGRATION=step2`.
  * DNS is swapped to the new stack.
  * Check that everything is functioning as expected.
3. `emp set EMPIRE_SCHEDULER_MIGRATION=step3`.
  * Old resources are removed.